### ISSUE TITLE
Let agents connect to control plane / api load balancer instead of first master ip

### ISSUE THREAD

**Issue by tkristiansen** [vitobotta/hetzner-k3s#61]
First of all, thanks for this nice project. I had some issues with my single master node today (it got very busy with both CPU and disk) and thought I'll upgrade to a HA cluster to not be so dependent on that single node. That worked fine - master 2 and master 3 got up running. The issues with master 1 did however come back, and I wondered why my whole cluster was then so irresponsive. It turns out that each of the agents are still connecting directly to master 1.

Are not the agents (and even the masters, maybe except nr1) supposed to connect to the API load balancer in order to not be dependent on a single master node?

**Comment by tkristiansen** [comment]
In this guide they also use the load balancer for internal communication: https://devopstales.github.io/kubernetes/k3s-etcd-kube-vip/ 

I haven't really verified it, but I assumed the agents would rely on the initial IP when (re-)connecting to the cluster.

**Comment by tkristiansen** [comment]
Btw., I'm not sure the issues I had today were actually related to this - I just stumbled across it and had this other guide that I had read before in mind.

I messed around with adding and removing master nodes and ended up doing a restore from backup (etcd snapshot) to the first master and then increased to three master nodes again. After that everything worked fine.

**Comment by vitobotta** [comment]
> In this guide they also use the load balancer for internal communication: https://devopstales.github.io/kubernetes/k3s-etcd-kube-vip/
> 
> I haven't really verified it, but I assumed the agents would rely on the initial IP when (re-)connecting to the cluster.

That's not a real load balancer, as described in that article.

**Comment by vitobotta** [comment]
> Btw., I'm not sure the issues I had today were actually related to this - I just stumbled across it and had this other guide that I had read before in mind.
> 
> I messed around with adding and removing master nodes and ended up doing a restore from backup (etcd snapshot) to the first master and then increased to three master nodes again. After that everything worked fine.

Glad to hear it's fixed. Di you have any problems with etcd backups? If you can share how you did it I will add notes to the README. I thought I had notes about it but I cannot find any.

**Comment by tkristiansen** [comment]
> That's not a real load balancer, as described in that article.

True, but as far as I understand, the purpose of using it internally is for the agent be able to reach any of the masters using one IP address. This would also be achieved when using the load balancer instead of a single master node. https://www.suse.com/c/rancher_blog/set-up-your-k3s-cluster-for-high-availability-on-digitalocean/ is also suggesting that the agents should connect to the load balancer.

But I guess the best would be to try it out - start up an HA cluster, take master 1 down and then restart one of the agents. If that works well, there should be no need for this change. I'll run a test and let you know how it goes.

> Di you have any problems with etcd backups? If you can share how you did it I will add notes to the README. I thought I had notes about it but I cannot find any.

I did try it twice before I got it to work, but cannot say in detail what I got wrong the first time. Basically, I followed the guide here: https://rancher.com/docs/k3s/latest/en/backup-restore/#restoring-a-cluster-from-a-snapshot

I'm not sure it was necessary, but I ended up rebuilding the whole `k3s server` command that you generate to initialize the master node and just replacing `--init-cluster` with these `--cluster-reset` and `--cluster-reset-restore-path` (the snapshot is by default stored in `${data-dir}/server/db/snapshots`).

This restore procedure will restore the snapshot and clear the knowledge about all other nodes. After that, you have to start the server again the normal way (I just ran `hetzner-k3s` again). Since the knowledge about the other master nodes are cleared, you should either clear etcd on any other master node as well, or just delete them and rebuild the cluster with the number of nodes you want and let them re-populate from master 1.

**Comment by vitobotta** [comment]
> True, but as far as I understand, the purpose of using it internally is for the agent be able to reach any of the masters using one IP address. This would also be achieved when using the load balancer instead of a single master node. https://www.suse.com/c/rancher_blog/set-up-your-k3s-cluster-for-high-availability-on-digitalocean/ is also suggesting that the agents should connect to the load balancer.

> But I guess the best would be to try it out - start up an HA cluster, take master 1 down and then restart one of the agents. If that works well, there should be no need for this change. I'll run a test and let you know how it goes.

I'm gonna change the code so that the other masters connect to the load balancer instead of the private IP of the first master. Let's see if it works without any issues.

> I'm not sure it was necessary, but I ended up rebuilding the whole k3s server command that you generate to initialize the master node and just replacing --init-cluster with these --cluster-reset and --cluster-reset-restore-path (the snapshot is by default stored in ${data-dir}/server/db/snapshots).

This is a candidate for another feature I could add, backup and restore. Will see if I can work on it when I have a bit more time.

**Comment by tkristiansen** [comment]
Ok, I guess changing to connect to the load balancer is good practice, but the agents do obviously cache or store the information about the other master nodes. I ran the test and this is the result:

I stopped master1 and worker1:

```
NAME                            STATUS     ROLES                       AGE     VERSION
temp-cpx21-master1              NotReady   control-plane,etcd,master   11m     v1.22.3+k3s1
temp-cpx21-master2              Ready      control-plane,etcd,master   10m     v1.22.3+k3s1
temp-cpx21-master3              Ready      control-plane,etcd,master   4m45s   v1.22.3+k3s1
temp-cpx21-pool-small-worker1   NotReady   <none>                      4m32s   v1.22.3+k3s1
temp-cpx21-pool-small-worker2   Ready      <none>                      4m32s   v1.22.3+k3s1
```

Then I started worker1:

```
NAME                            STATUS     ROLES                       AGE     VERSION
temp-cpx21-master1              NotReady   control-plane,etcd,master   11m     v1.22.3+k3s1
temp-cpx21-master2              Ready      control-plane,etcd,master   10m     v1.22.3+k3s1
temp-cpx21-master3              Ready      control-plane,etcd,master   4m48s   v1.22.3+k3s1
temp-cpx21-pool-small-worker1   Ready      <none>                      4m35s   v1.22.3+k3s1
temp-cpx21-pool-small-worker2   Ready      <none>                      4m35s   v1.22.3+k3s1
```

As you see, it did recover fine even when master1 was not running.

**Comment by vitobotta** [comment]
Awesome. I am testing the load balancer change right now.

**Comment by tkristiansen** [comment]
So the above test was for the agent. Now I tested with the masters. After taking down master3, I get `Error from server (InternalError): an error on the server ("apiserver not ready") has prevented the request from succeeding (get nodes)`. The reason is that we don't have a majority of the masters available and it cannot rely on the single master we have left.

As soon as I start up master3 again (master1 still left off), I get this:

```
NAME                            STATUS     ROLES                       AGE   VERSION
temp-cpx21-master1              NotReady   control-plane,etcd,master   17m   v1.22.3+k3s1
temp-cpx21-master2              Ready      control-plane,etcd,master   16m   v1.22.3+k3s1
temp-cpx21-master3              NotReady   control-plane,etcd,master   10m   v1.22.3+k3s1
temp-cpx21-pool-small-worker1   Ready      <none>                      10m   v1.22.3+k3s1
temp-cpx21-pool-small-worker2   Ready      <none>                      10m   v1.22.3+k3s1
```

which after a while turns into:

```
NAME                            STATUS     ROLES                       AGE   VERSION
temp-cpx21-master1              NotReady   control-plane,etcd,master   19m   v1.22.3+k3s1
temp-cpx21-master2              Ready      control-plane,etcd,master   19m   v1.22.3+k3s1
temp-cpx21-master3              Ready      control-plane,etcd,master   13m   v1.22.3+k3s1
temp-cpx21-pool-small-worker1   Ready      <none>                      12m   v1.22.3+k3s1
temp-cpx21-pool-small-worker2   Ready      <none>                      12m   v1.22.3+k3s1
```

As far as I understand, the availability between the masters relies solely on the etcd / raft protocol which will have to know which server it is actually talking with.

Again, this shows that the way you originally had it should work fine, but let's hope that using the load balancer IP internally is at least not breaking anything - based on what others are doing, it should be fine :-)

**Comment by vitobotta** [comment]
I just released 0.5.2. Now when the additional masters join the first one connect to the load balancer for the API server. I have been able to confirm that with the first master down I could still spin up workloads etc. So this means that indeed the load balancer is now used internally.

**Comment by vitobotta** [comment]
Can you please try a cluster with 0.5.2 and confirm that you can still do stuff with master 1 down?

**Comment by tkristiansen** [comment]
Confimed. Looks like you changed the kubeconfig file as well to use the load balancer IP. Now I was able to interact with the cluster while master1 is down without having to do any manual adjustments at all.

Also verified that taking down another master will make the cluster unavailable, and an agent will recover after restart (same tests as I did before with the same expected behavior).

Thanks for this improvement!
